src/alerts.rs

```diff
@@ -159,47 +236,141 @@ impl<H: Hasher, D: Data, KB: KeyBox> Alerter<H, D, KB> {
         Some(full_unit1.creator())
     }
 
-    fn on_own_alert(&mut self, alert: Alert<H, D, KB::Signature>) {
-        if let Err(e) = self
-            .messages_for_network
-            .unbounded_send((AlertMessage::ForkAlert(alert), Recipient::Everyone))
-        {
-            debug!(target: "alerter", "Error when broadcasting units {:?}.", e);
+    fn rmc_alert(&mut self, forker: NodeIndex, alert: Signed<'a, Alert<H, D, MK::Signature>, MK>) {
+        let hash = alert.as_signable().hash();
+        self.known_rmcs
+            .insert((alert.as_signable().sender, forker), hash);
+        self.known_alerts.insert(hash, alert);
+        self.rmc.start_rmc(hash);
+    }
+
+    fn on_own_alert(&mut self, alert: Alert<H, D, MK::Signature>) {
+        let forker = alert.proof.0.as_signable().creator();
+        self.known_forkers.insert(forker, alert.proof.clone());
+        let alert = Signed::sign(alert, self.keychain);
+        if let Err(e) = self.messages_for_network.unbounded_send((
+            AlertMessage::ForkAlert(alert.clone().into()),
+            Recipient::Everyone,
+        )) {
+            debug!(target: "alerter", "{:?} Error when broadcasting alert {:?}.", self.index(), e);
         }
+        self.rmc_alert(forker, alert);
     }
 
-    fn on_network_alert(&mut self, alert: Alert<H, D, KB::Signature>) {
-        if let Some(forker) = self.who_is_forking(&alert.proof) {
+    fn on_network_alert(
+        &mut self,
+        alert: UncheckedSigned<Alert<H, D, MK::Signature>, MK::Signature>,
+    ) {
+        let alert = match alert.check(self.keychain) {
+            Ok(alert) => alert,
+            Err(e) => {
+                debug!(target: "alerter","{:?} We have received an incorrectly signed alert: {:?}.", self.index(), e);
+                return;
+            }
+        };
+        let contents = alert.as_signable();
+        if let Some(forker) = self.who_is_forking(&contents.proof) {
+            if self.known_rmcs.contains_key(&(contents.sender, forker)) {
+                debug!(target: "alerter","{:?} We already know about an alert by {:?} about {:?}.", self.index(), alert.as_signable().sender, forker);
```

Comment:
What will happen if we initially get some alert A for the pair (i, j) (and we were the only node to receive A) and all the other nodes received A' and actually completed RMC on A'?

Reply:
O, nice catch, this is definitely a bug. Will require a bit of work to fix.